### ISSUE TITLE
Heap corruption of Lists of numpy arrays

### ISSUE THREAD

**Issue by Viissataa** [numba/numba#5264]
Lists of numpy arrays (at least of float 1D type) cannot be brought to length zero without heap corruption. Best guess at this point: refcount issue and double free of allocated memory.

Also, Lists of numpy scalars for some reason do not suffer this issue.

Minimal reproducer:

```python
import numpy as np
from numba import  float64
from numba.typed import List

float_array = float64[:]

l = List.empty_list(float_array)
l.append(np.ones(3,dtype=np.float64))
l.pop()
````

**Comment by esc** [comment]
@Viissataa thanks for reporting this on the Numba issue tracker. I have reproduced this locally and can confirm, that it is likely a bug.

**Comment by stuartarchibald** [comment]
After a bit of debugging, this is likely a refcounting issue, essentially a double free is the root cause of the SIGSEGV.

**Comment by esc** [comment]
@Viissataa sorry for the delay on this one, I just got around to looking at it. I changed the reproducer slightly, so as to assign the incoming array to a variable. I then no longer receive the segfault. I was wondering if you could try the example and see if you see the same?

```
import numpy as np
from numba import float64
from numba.typed import List

float_array = float64[:]

l = List.empty_list(float_array)
z = np.ones(3,dtype=np.float64)
l.append(z)
l.pop()
print(repr(l))
```

**Comment by Viissataa** [comment]
I still get segfault
![2020-04-20 17 43 07](https://user-images.githubusercontent.com/7416056/79764886-9df47780-832e-11ea-9c9e-4e8a221ab88c.jpg)

My numba version is 0.46.0 as per pip.

**Comment by esc** [comment]
Possible fix in #5602 - @Viissataa would you have a chance to try that fix?
